### ISSUE TITLE
Recommended procedure for taking measurements?

### ISSUE THREAD

**Issue by luisCartoGeo** [facebookresearch/map-anything#35]
Hello, thank you very much for your great responsiveness. It is an honor to be testing your model from Latin America. I have a question, I hope you can guide me.

I want to measure the distances, widths, and heights of objects in the images.

To do this, I try to overlay pts3d on each output image with their respective intrinsics generated by the model.
I observe good alignment in two images; however, in one of them, the points extend beyond the range.
Nevertheless, the representation of the points in 3D shows a good representation in all images.
I am attaching the images for your consideration.

I will ask the questions and then provide the details, including images and code:

1. Is it possible to perform metric calculations by calculating the Euclidean distance of pts3d?
2. Is it necessary to calculate a scale factor on the X and Y axes to obtain better results?
3. If I use several images with pts3d, which one is recommended for performing these calculations or the recommended preliminary procedure?
4. Where am I going wrong in my procedure for superimposing the images and pts3d to perform calculations? I have attached images and code.

**Results when plotting in 2D and 3D:**

<img width="1024" height="568" alt="Image" src="https://github.com/user-attachments/assets/fcb70b37-14c9-4e34-bb21-8aa0297f68a2" />

**2D graph**:
I use Plotly Dash to perform interactive distance calculations in Colab.

**Python code for the 2D graph**
```
import plotly.graph_objects as go
import numpy as np
import PIL
from PIL import Image, ImageEnhance, ImageOps
import io
import base64
import torch
import dash
from dash import dcc, html, Input, Output

app = dash.Dash(__name__)

dp=[]

#-------------ENTRADA: IMAGENES, PUNTOS y K--------------------
ie=imgsf[0]
pe=pts[0]
K = intrinc[0]

# === 1. Imagen de pred['img_no_norm'] ===
#img = pred['img_no_norm'][0].cpu().numpy()
img =ie[0].cpu().numpy()

# Normalizar si es float [0,1]
if img.max() <= 1.0:
    img = (img * 255).clip(0,255).astype(np.uint8)
else:
    img = img.astype(np.uint8)


# Convertir a base64 para Plotly
img_pil = PIL.Image.fromarray(img)

# Mejorar brillo (factor >1 = más brillo, <1 = más oscuro)
enhancer_brightness = ImageEnhance.Brightness(img_pil)
img_brillo = enhancer_brightness.enhance(2.5)  # 1.5 = 50% más brillante
# Mejorar contraste (factor >1 aumenta, <1 disminuye)
#enhancer = ImageEnhance.Contrast(img_brillo)
#img_contraste = enhancer.enhance(1.2)  # duplica contraste

#otra opcion de contraste automatico
img_mejorada = ImageOps.autocontrast(img_brillo, cutoff=0.25)

buffer = io.BytesIO()
img_pil.save(buffer, format="PNG")
img_str = "data:image/png;base64," + base64.b64encode(buffer.getvalue()).decode()

# === 2. Puntos 3D ===
pts3d1 = pe.float()
if pts3d1.ndim == 4:
    pts3d1 = pts3d1.squeeze(0)
pts3d_flat = pts3d1.reshape(-1, 3)

# Filtrar Z>0
mask = pts3d_flat[:, 2] > 0
pts3d_flat = pts3d_flat[mask]

# Proyección 3D -> 2D
pts3d_h = torch.cat([pts3d_flat, torch.ones((pts3d_flat.shape[0], 1))], dim=1).T
pts2d_flat = K @ pts3d_h[:3]
pts2d_flat = pts2d_flat[:2] / pts2d_flat[2]
pts2d_flat = pts2d_flat.T.cpu().numpy()

# === 3. Hover con XYZ y norma ===
hover_text = [
    f"X: {x:.2f}<br>Y: {y:.2f}<br>Z: {z:.2f}<br>|P|: {np.sqrt(x**2+y**2+z**2):.2f}"
    for x,y,z in pts3d_flat.numpy()
]

# === 4. Figura Plotly ===
fig = go.Figure()

# Imagen de fondo
fig.add_layout_image(
    dict(
        source=img_str,
        xref="x", yref="y",
        x=0, y=0,
        sizex=img.shape[1], sizey=img.shape[0],
        sizing="contain", opacity=1,
        layer="below"
    )
)

# Puntos proyectados
fig.add_trace(go.Scatter(
    x=pts2d_flat[:, 0],
    y=pts2d_flat[:, 1],
    mode='markers',
    marker=dict(color='red', size=0.5, opacity=0.9),
    text=hover_text,
    hoverinfo="text"
))

# Ajustes de ejes
fig.update_yaxes(autorange="reversed")
fig.update_xaxes(range=[0, img.shape[1]],showgrid=False)
fig.update_yaxes(range=[0, img.shape[0]],showgrid=False)
fig.update_layout(width=1000, height=800, title="Distancias 3D (Hover con |P|)")

app.layout = html.Div([
    dcc.Graph(id="grafico", figure=fig),
    html.Div(id="output")
])

@app.callback(
    Output("output", "children"),
    Input("grafico", "clickData")
)
def mostrar(clickData):
    if clickData:
        punto = clickData["points"][0]
        t = punto['text']
        x=float(t[t.find('X')+2:t.find('<')])
        t=t[t.find('>')+1:]
        y= float(t[t.find('Y')+2:t.find('<')])
        z= float(t[t.find('Z')+2:t.rfind('<')])

        if len(dp)==0:
           dp.append([x,y,z])
           return f"Guardado: {t}"
        else:
           p1=np.array(dp[0])
           p2=np.array([x,y,z])
           #distancia euclideana
           dist = np.linalg.norm(np.array(p2) - np.array(p1))
           dp.clear()
           return f"Guardado: {dist}"
    return "Haz clic en un punto"

if __name__ == "__main__":
    app.run(debug=True)
```

**Modification of points to apply the pose**
I also tried applying the appropriate pose to each image, checking beforehand whether it needed to be inverted, but this did not work either.

```
# === Transformar al sistema de la cámara ===
pose = pose[0].float()   # i = índice de la vista

pts3d_cam = (pose @ pts3d_h).T[:, :3]

#verificar cual es la pose correcta, es pose sin invertir 
#pts3d_cam_inv = (pose_inv @ pts3d_h).T[:, :3]  # opción A
#pts3d_cam_dir = (pose    @ pts3d_h).T[:, :3]  # opción B

#print("Z>0 con pose_inv:", (pts3d_cam_inv[:,2]>0).float().mean().item())
#print("Z>0 con pose    :", (pts3d_cam_dir[:,2]>0).float().mean().item())

# Filtrar Z>0 solo los puntos frente a la camara
mask = pts3d_cam[:, 2] > 0
pts3d_cam = pts3d_cam[mask]

# Proyección con K
pts3d_h_cam = torch.cat([pts3d_cam, torch.ones((pts3d_cam.shape[0], 1))], dim=1).T

K = torch.tensor(K).float() #no se aplicaba antes y funcionaba

pts2d_flat = K @ pts3d_h_cam[:3]
pts2d_flat = pts2d_flat[:2] / pts2d_flat[2]
pts2d_flat = pts2d_flat.T.cpu().numpy()
```

**Comment by luisCartoGeo** [comment]
Using points from the camera perspective (pts3d_cam) seems to solve the alignment problem.
The other questions remain:

1. Recommended procedure for taking measurements?
2. Is it possible to perform metric calculations by calculating the Euclidean distance of pts3d?
4. Is it necessary to calculate a scale factor on the X and Y axes to obtain better results?
5. In the example described, I have found differences in measurements of +/- 20 cm, especially on the Y axis, which is smaller than the X axis. Is it possible to help the model by introducing a distance of known points? Or have you considered whether such an option would be possible?
6. If I use several images with pts3d, which one is recommended for performing these calculations or the recommended preliminary procedure?

**Comment by Nik-V9** [comment]
Hi, thanks for testing MapAnything on your applications!

As you mentioned, I would recommend using `pts3d_cam` or recomputing points from `depth_z` and `intrinsics` for the best alignment to the image.

Answer to your questions:

1. The euclidean distance of any two points in `pts3d_cam` can be used for measurement: https://github.com/facebookresearch/map-anything/blob/5ffcb617601a40687e0daf06cd625959adeb4472/scripts/gradio_app.py#L720
2. Yes
3. I'm not sure I follow this question. Regardless, there should be no external scaling required
4. The metric scale estimation from the model would not be perfect. However, as you mentioned, you can provide the distance of known points (represented as sparse depth) to the model and this should improve the metric scale estimation.
5. I believe that all images should give you similar quality. I would recommend using the images and pts3d_cam that best suit your use case.
